### ISSUE TITLE
Row.get seems to mix up columns

### ISSUE THREAD

**Issue by schauder** [r2dbc/r2dbc-postgresql#86]
I'm preparing an example for a conference talk.

I'm storing 1703 rows of data in a simple table
```
CREATE TABLE ROW( 
    ID SERIAL PRIMARY KEY, 
    CONTENT VARCHAR(2000)
)
```

When I try to load it again with: 
```
RowRepository rows;
\\....
    rows.findAll();
```
I get some rows back (typically ~30 sometimes up to ~150) before I get an:

```
java.lang.NumberFormatException: For input string: ""
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
```
(Full stack trace is below)

The fun part is that the string quoted seems most of the time two non-printable characters (boxes in the IDE console) but sometimes small bits of text like `"ink"` which leaves me with the impression that somehow data that belongs in one column ends up in another column.

```
org.springframework.data.mapping.MappingException: Couldn't read column id from Row.
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter$RowParameterValueProvider.getParameterValue(MappingR2dbcConverter.java:326) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.data.relational.core.conversion.BasicRelationalConverter$ConvertingParameterValueProvider.getParameterValue(BasicRelationalConverter.java:256) ~[spring-data-relational-1.1.0.BUILD-20190312.114323-483.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.data.convert.ClassGeneratingEntityInstantiator$EntityInstantiatorAdapter.extractInvocationArguments(ClassGeneratingEntityInstantiator.java:250) ~[spring-data-commons-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.data.convert.ClassGeneratingEntityInstantiator$EntityInstantiatorAdapter.createInstance(ClassGeneratingEntityInstantiator.java:223) ~[spring-data-commons-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.data.convert.ClassGeneratingEntityInstantiator.createInstance(ClassGeneratingEntityInstantiator.java:84) ~[spring-data-commons-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.data.relational.core.conversion.BasicRelationalConverter.createInstance(BasicRelationalConverter.java:146) ~[spring-data-relational-1.1.0.BUILD-20190312.114323-483.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.createInstance(MappingR2dbcConverter.java:174) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:103) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:98) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:56) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:46) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:29) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	at io.r2dbc.postgresql.PostgresqlResult.lambda$map$0(PostgresqlResult.java:69) ~[r2dbc-postgresql-1.0.0.M7.jar:na]
	at reactor.function.TupleUtils.lambda$function$7(TupleUtils.java:143) ~[reactor-extra-3.2.2.RELEASE.jar:3.2.2.RELEASE]
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:100) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxZip$ZipCoordinator.drain(FluxZip.java:735) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxZip$ZipCoordinator.request(FluxZip.java:587) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:155) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxFlatMap$FlatMapInner.request(FluxFlatMap.java:965) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:681) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:540) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.request(FluxFlatMap.java:311) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.Operators$DeferredSubscription.request(Operators.java:1364) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:1842) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:388) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.innerComplete(FluxConcatMap.java:289) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onComplete(FluxConcatMap.java:873) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1508) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.signal(MonoDelayUntil.java:211) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.MonoDelayUntil$DelayUntilTrigger.onComplete(MonoDelayUntil.java:290) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:118) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_152]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_152]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_152]
Caused by: java.lang.NumberFormatException: For input string: ""
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_152]
	at java.lang.Integer.parseInt(Integer.java:569) ~[na:1.8.0_152]
	at java.lang.Integer.parseInt(Integer.java:615) ~[na:1.8.0_152]
	at io.r2dbc.postgresql.codec.IntegerCodec.doDecode(IntegerCodec.java:60) ~[r2dbc-postgresql-1.0.0.M7.jar:na]
	at io.r2dbc.postgresql.codec.IntegerCodec.doDecode(IntegerCodec.java:31) ~[r2dbc-postgresql-1.0.0.M7.jar:na]
	at io.r2dbc.postgresql.codec.AbstractCodec.decode(AbstractCodec.java:64) ~[r2dbc-postgresql-1.0.0.M7.jar:na]
	at io.r2dbc.postgresql.codec.DefaultCodecs.decode(DefaultCodecs.java:92) ~[r2dbc-postgresql-1.0.0.M7.jar:na]
	at io.r2dbc.postgresql.PostgresqlRow.get(PostgresqlRow.java:83) ~[r2dbc-postgresql-1.0.0.M7.jar:na]
	at io.r2dbc.spi.Row.get(Row.java:46) ~[r2dbc-spi-1.0.0.M7.jar:na]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter$RowParameterValueProvider.getParameterValue(MappingR2dbcConverter.java:324) ~[spring-data-r2dbc-1.0.0.BUILD-20190310.103734-31.jar:1.0.0.BUILD-SNAPSHOT]
	... 38 common frames omitted
```

This is the commit that reproduces the problem:
https://github.com/schauder/talk-reactive-relational-db/commit/29bee84d4193b0c5fd655391953ef18d15bd6655

I'm using:
A local snapshot build of spring-data-r2dbc a few days old.
r2dbc-postgresql 1.0.0.M7

**Comment by mp911de** [comment]
I wonder whether Spring Data R2DBC tries to read the value twice. Last time I checked, it was possible to read a value only once, the second attempt failed because the backing buffer was exhausted (read: readIndex() was not reset).

**Comment by croudet** [comment]
I had the same problem when I tried to use r2dbc-postgresql with micronaut (https://micronaut.io/).

R2bc was able to retrieve couples of rows but then the buffer index will wrong and will get the same exception.

**Comment by nebhale** [comment]
@schauder Sorry for the delay in getting back to you on this.  I just tried to reproduce the problem against the driver directly (without Spring Data involved) and couldn't get the behavior you're describing.  Take a look at [my Gist](https://gist.github.com/nebhale/e75f426a1406461410c8ba99f550cdfd) and let me know if I've done something incorrect.

**Comment by croudet** [comment]
On my end, I have something like that:
```java
private @Inject @Named("dsR2DBC") R2dbc dsR2dbc;

@Override
public Flux<Asset> listAssets(Integer offset, Integer max, SortOrder sort) {
        return dsR2dbc.inTransaction(handle -> handle.createQuery(listAssetsQuery(offset, max, sort)).mapRow(ASSET_RESULT_MAPPER_WITHOUT_EXTRA_FIELDS));
}
```
And if I use such a subscriber (`Subsciption.request` in antoher thread):
```java
class MySubscriber implements Subscriber<Asset> {
        Subscription s;

        @Override
        public void onSubscribe(Subscription s) {
            this.s = s;
            s.request(1L);
        }

        @Override
        public void onNext(Asset t) {
            new Thread(() -> s.request(1L)).start();
        }

        @Override
        public void onError(Throwable t) {
        }

        @Override
        public void onComplete() {
        }

    }
```
I get:
```
onError java.lang.StringIndexOutOfBoundsException: offset 16967, count 5, length 0 Thread[reactor-tcp-nio-1,5,main]
java.lang.StringIndexOutOfBoundsException: offset 16967, count 5, length 0
        at java.base/java.lang.String.checkBoundsOffCount(String.java:3395)
        at java.base/java.lang.String.<init>(String.java:514)
        at io.netty.buffer.ByteBufUtil.decodeString(ByteBufUtil.java:797)
        at io.netty.buffer.AbstractByteBuf.toString(AbstractByteBuf.java:1239)
        at io.netty.buffer.AbstractByteBuf.getCharSequence(AbstractByteBuf.java:504)
        at io.netty.buffer.AbstractUnpooledSlicedByteBuf.getCharSequence(AbstractUnpooledSlicedByteBuf.java:264)
        at io.netty.buffer.AbstractByteBuf.readCharSequence(AbstractByteBuf.java:509)
        at io.r2dbc.postgresql.util.ByteBufUtils.decode(ByteBufUtils.java:42)
        at io.r2dbc.postgresql.codec.LongCodec.doDecode(LongCodec.java:60)
        at io.r2dbc.postgresql.codec.LongCodec.doDecode(LongCodec.java:31)
        at io.r2dbc.postgresql.codec.AbstractCodec.decode(AbstractCodec.java:64)
        at io.r2dbc.postgresql.codec.DefaultCodecs.decode(DefaultCodecs.java:92)
        at io.r2dbc.postgresql.PostgresqlRow.get(PostgresqlRow.java:83)
```

**Comment by schauder** [comment]
@nebhale the Gist seems fine.

So this might be a Spring Data issue. Is there a way to move this issue over to Spring Data R2DBC?

**Comment by mp911de** [comment]
This isn't a Spring Data issue. It's related to the context of how we access Postgres. I'm not able to reproduce the issue in a standalone-gist. Running the exact same code with WebFlux and SSE (this is the missing bit to reproduce the issue) causes for me reliably `java.lang.NumberFormatException: For input string: ""`.

**Comment by schauder** [comment]
I wonder if it is a concurrency issue. Maybe flatmapping/grouping the result makes it possible to reproduce it?

**Comment by mp911de** [comment]
Disabling SSE caused the issue to disappear. I do not know how/why SSE interferes here. We're not emitting bare `DataBuffer`s through `Publisher` so disposal/buffer release isn't interfering.

Paging @rstoyanchev for fyi/cc.

**Comment by schauder** [comment]
SSE?

**Comment by mp911de** [comment]
Your code example uses Server-Sent-Events (`@GetMapping(value = "/all", produces = MediaType.TEXT_EVENT_STREAM_VALUE)`). Switching to a plain list (i.e. removing `produces = …`) causes the issue to disappear.

**Comment by croudet** [comment]
I have this gist: https://gist.github.com/croudet/e34275f758765f7799b4870a45e8163d.js

This gives me:
```
java.lang.NumberFormatException: For input string: "ue"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:68)
	at java.base/java.lang.Integer.parseInt(Integer.java:658)
	at java.base/java.lang.Integer.parseInt(Integer.java:776)
	at io.r2dbc.postgresql.codec.IntegerCodec.doDecode(IntegerCodec.java:60)
	at io.r2dbc.postgresql.codec.IntegerCodec.doDecode(IntegerCodec.java:31)
	at io.r2dbc.postgresql.codec.AbstractCodec.decode(AbstractCodec.java:64)
	at io.r2dbc.postgresql.codec.DefaultCodecs.decode(DefaultCodecs.java:92)
	at io.r2dbc.postgresql.PostgresqlRow.get(PostgresqlRow.java:83)
	at net.activia.diffserver.ColumnMixUpTest.lambda$columnMixUp$0(ColumnMixUpTest.java:65)
	at io.r2dbc.postgresql.PostgresqlResult.lambda$map$0(PostgresqlResult.java:69)
	at reactor.function.TupleUtils.lambda$function$7(TupleUtils.java:143)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:100)
	at reactor.core.publisher.FluxZip$ZipCoordinator.drain(FluxZip.java:735)
	at reactor.core.publisher.FluxZip$ZipCoordinator.request(FluxZip.java:587)
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:155)
	at reactor.core.publisher.FluxFlatMap$FlatMapInner.request(FluxFlatMap.java:965)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:681)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:540)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.request(FluxFlatMap.java:311)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:1842)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.request(MonoFlatMapMany.java:102)
	at reactor.core.publisher.StrictSubscriber.request(StrictSubscriber.java:138)
	at ColumnMixUpTest$MySubscriber.lambda$onNext$0(ColumnMixUpTest.java:101)
	at java.base/java.lang.Thread.run(Thread.java:835)
```

**Comment by mp911de** [comment]
Thanks a lot @croudet. I also suspected threads when using SSE that they cause some interference.

Out of curiosity: @croudet can you build the same test for SQL server? We should expect the same issue happening there.

**Comment by croudet** [comment]
I am sorry I don't have SQL server.

**Comment by mp911de** [comment]
Did a cursory test on SQL server. Wasn't able to reproduce the issue with a threaded `request(n)` in `r2dbc-mssql`.

**Comment by mp911de** [comment]
That's fixed now via #116.
